### ISSUE TITLE
Race condition on saving Models/indexing to ES

### ISSUE THREAD

**Issue by Rafalonso** [sabricot/django-elasticsearch-dsl#231]
When updating a field on a model twice at the same time may lead to unmatched values:

```python
def update_django_model(name):
    DjangoModelObject = DjangoModel.objects.first()
    DjangoModelObject.name = name
    DjangoModelObject.save()

p = Pool(2)
p.map(update_django_model, ['first', 'second'])
```

This may lead to the db having the value "second" and the ES object having "first" on the `name` field.
Solving this problem may also solve https://github.com/sabricot/django-elasticsearch-dsl/issues/96
Either using the same transaction for each model update or locking the table on these ids are good solutions

**Comment by safwanrahman** [comment]
I think django ORM is not good suitable for asynchronous code. With django 3.0, your code may not work.
But I understand the situation, as both of the process runing parallely, how do you ensure that the `second` is gone to Database before `first`?

**Comment by Rafalonso** [comment]
With the code I demonstrated I'm not concerned with the order in which these fields get saved but the sync between ES and DB. `first` on both or `second` on both.

Since the library uses signals after any transaction has been committed it's hard to make sure that the signal executes immediately after the save.

**Comment by safwanrahman** [comment]
@Rafalonso That is expected actually. It is a problem that can not be solved from the application end (I think). As Elasticsearch does not support transaction, we need to insert the data after the commit is saved to database, I do not know if there are any other way to keep it sync with the database. If you need this kind of sync, I think it is better to use some kind of DB plugin like [zomobodb](https://www.zombodb.com/) for postgresql.

**Comment by safwanrahman** [comment]
@Rafalonso Do you have any idea about how this can be solved?

**Comment by Rafalonso** [comment]
I'm exploring both in-application layer solutions as well as solutions out of this library's scope.

So far manually locking the row on pre_save() signal looks promising, I'll report back with results.

**Comment by safwanrahman** [comment]
@Rafalonso If you update the ES in `pre_save`, you are not sure about whether the transaction actually got saved. As elasticsearch does not support transactions, it is not possible.

**Comment by Rafalonso** [comment]
That's true but if you lock the database in pre-save and then release in post_save after indexing, that problem shouldn't arise

**Comment by safwanrahman** [comment]
That will be much more complicated from django end and not every database backend support DB locking, moreover it will degregade the performance. I think it is better to use DB level plugin for having this level of concurrency.

**Comment by Rafalonso** [comment]
Indeed will be hard to maintain support for all DB backends. This is very specific depending on the application. Since `post_save` indexes outside a database transaction I believe that this problem is outside the purview of the library.
This issue can be closed or be left open for a few more days to see if anyone can come up with other solutions. Up to you ;)

Thank you for your comments.

**Comment by alexgarel** [comment]
@Rafalonso I think actual implementation in library is simple. 
If you have less tolerance on coherence, I  would suggest, not using the signal as is, but:

- if you use Postgresqql, use on_commit to trigger the code equivalent to the signal
- always reread object from database before indexation
